File: samples-scala/src/test/scala/io/github/ilaborie/loops/samples/FilterTest.scala
package io.github.ilaborie.loops.samples

import org.junit.Assert.assertEquals
import org.junit.Test

class FilterTest {


  private def test(func: List[ElementScala] => List[ElementScala]): Unit = {
    val elements = List(
      ElementScala("a"),
      ElementScala("b"),
      ElementScala("aa")
    )

    val result = func(elements)

    assertEquals(List(ElementScala("a"), ElementScala("aa")), result)

  }

  @Test
  def for_should_filter(): Unit = {
    test {
      io.github.ilaborie.loops.samples.anatomy.FilterScala.filter
    }
  }

  @Test
  def recursion_should_filter(): Unit = {
    test {
      io.github.ilaborie.loops.samples.recursion.FilterScala.filter
    }
  }

  @Test
  def recursion_should_filter_alt(): Unit = {
    test {
      io.github.ilaborie.loops.samples.recursion.FilterScala.filterAlt
    }
  }

  @Test
  def tail_recursion_should_filter(): Unit = {
    test {
      io.github.ilaborie.loops.samples.recursion.FilterTrScala.filter
    }
  }

  @Test
  def tail_recursion_should_filter_alt(): Unit = {
    test {
      io.github.ilaborie.loops.samples.recursion.FilterTrScala.filterAlt
    }
  }

  @Test
  def stream_should_filter(): Unit = {
    test {
      io.github.ilaborie.loops.samples.stream.FilterScala.filter
    }
  }

}

File: samples-kotlin/src/test/kotlin/io/github/ilaborie/loops/samples/stream/FilterTest.kt
package io.github.ilaborie.loops.samples.stream

import io.github.ilaborie.loops.samples.ElementKt
import org.junit.Assert.assertEquals
import org.junit.Test

class FilterTest {

    private fun test(func: (List<ElementKt>) -> List<ElementKt>) {
        val elements = listOf(
            ElementKt("a"),
            ElementKt("b"),
            ElementKt("aa")
        )

        val result = func(elements)

        assertEquals(listOf(ElementKt("a"), ElementKt("aa")), result)
    }


    @Test
    fun for_should_filter() {
        test { io.github.ilaborie.loops.samples.anatomy.filter(it) }
    }

    @Test
    fun recursion_should_filter() {
        test { io.github.ilaborie.loops.samples.recursion.filter(it) }
    }

    @Test
    fun recursion_should_filter_alt() {
        test { io.github.ilaborie.loops.samples.recursion.filterAlt(it) }
    }

    @Test
    fun tail_recursion_should_filter() {
        test { io.github.ilaborie.loops.samples.recursion.filterTr(it) }
    }

    @Test
    fun tail_recursion_should_filter_alt() {
        test { io.github.ilaborie.loops.samples.recursion.filterTrAlt(it) }
    }

    @Test
    fun stream_should_filter() {
        test { io.github.ilaborie.loops.samples.stream.filter(it) }
    }
}
